### PR TITLE
Add externally managed mode API

### PR DESCRIPTION
**Pull request by SeanFeldman**
To create externally managed container endpoint using MS DI w/o exposing the internals

### CODE REVIEW

**Comment by SeanFeldman** [comment]
Need a test that verifies this for all the supported containers

File: src/NServiceBus.Extensions.DependencyInjection/IStartableEndpointWithExternallyManagedContainerExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    public static class IStartableEndpointWithExternallyManagedContainerExtensions
+    {
+        /// <summary>
+        /// Starts the endpoint and returns a reference to it.
+        /// </summary>
+        /// <param name="startableEndpoint">The startable endpoint.</param>
+        /// <param name="serviceProvider">The container specific service provider factory.</param>
+        /// <returns>A reference to the endpoint.</returns>
+        public static Task<IEndpointInstance> Start(this IStartableEndpointWithExternallyManagedContainer startableEndpoint, IServiceProvider serviceProvider)
+        {
```
Comment (timbussmann) [inline src/NServiceBus.Extensions.DependencyInjection/IStartableEndpointWithExternallyManagedContainerExtensions.cs]: should we add the guards here too?

File: src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+﻿namespace NServiceBus
+{
+    using Extensions.DependencyInjection;
+    using Microsoft.Extensions.DependencyInjection;
+
+    /// <summary>
+    /// Provides factory methods for creating endpoints instances with an externally managed container implementing Microsoft dependency injection abstraction.
+    /// </summary>
+    public static class EndpointWithExternallyManagedServiceProvider
```
Comment (timbussmann) [inline src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs]: can't come up with a better name and it's technically correct so I'm 👍

Comment (andreasohlund) [reply src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs]: Agree 👍

File: src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        /// <summary>
+        /// Creates a new startable endpoint based on the provided configuration that uses an externally managed container implementing Microsoft dependency injection abstraction.
+        /// </summary>
+        /// <param name="endpointConfiguration">The endpoint configuration.</param>
+        /// <param name="serviceCollection">Service collection</param>
```
Comment (andreasohlund) [inline src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs]:
```suggestion
        /// <param name="serviceCollection">Service collection.</param>
```

File: src/NServiceBus.Extensions.DependencyInjection/Guard.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿namespace NServiceBus.Extensions.DependencyInjection
```
Comment (andreasohlund) [inline src/NServiceBus.Extensions.DependencyInjection/Guard.cs]:
```suggestion
﻿namespace NServiceBus
```

See https://github.com/Particular/PlatformArchitecture/blob/master/principles/cross-cutting-concepts.md#nservicebus-rules

File: src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+﻿namespace NServiceBus
+{
+    using Extensions.DependencyInjection;
+    using Microsoft.Extensions.DependencyInjection;
+
+    /// <summary>
+    /// Provides factory methods for creating endpoints instances with an externally managed container implementing Microsoft dependency injection abstraction.
```
Comment (andreasohlund) [inline src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs]:
```suggestion
    /// Provides factory methods for creating endpoints instances with an externally managed container implementing the Microsoft dependency injection abstraction.
```

File: src/NServiceBus.Extensions.DependencyInjection/IStartableEndpointWithExternallyManagedContainerExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Extensions.DependencyInjection;
+
+    /// <summary>
+    /// An extension point for to create an endpoint in the start-up phase where the container is externally managed using Microsoft dependency injection abstraction.
```
Comment (andreasohlund) [inline src/NServiceBus.Extensions.DependencyInjection/IStartableEndpointWithExternallyManagedContainerExtensions.cs]:
```suggestion
    /// An extension point for to create an endpoint in the start-up phase where the container is externally managed using the Microsoft dependency injection abstraction.
```

File: src/NServiceBus.Extensions.DependencyInjection.Autofac.AcceptanceTests/When_using_externally_managed_container.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        [Test]
+        public async Task Should_resolve_dependencies_from_serviceprovider()
+        {
+            IContainer container = null;
+            var context = await Scenario.Define<Context>()
+                .WithEndpoint<EndpointWithExternallyManagedContainer>(endpointBuilder =>
+                {
+                    var serviceCollection = new ServiceCollection();
+                    endpointBuilder.CustomConfig(endpointConfiguration => endpointConfiguration
```
Comment (andreasohlund) [inline src/NServiceBus.Extensions.DependencyInjection.Autofac.AcceptanceTests/When_using_externally_managed_container.cs]: Should we add a component via the servicecollection as well?

File: src/NServiceBus.Extensions.DependencyInjection.Autofac.AcceptanceTests/When_using_externally_managed_container.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        {
+        }
+
+        class InternalService
+        {
+        }
+
+        class ExternalService
+        {
+
```
Comment (andreasohlund) [inline src/NServiceBus.Extensions.DependencyInjection.Autofac.AcceptanceTests/When_using_externally_managed_container.cs]:
```suggestion
```

File: src/NServiceBus.Extensions.DependencyInjection.Autofac.AcceptanceTests/When_using_externally_managed_container.cs (hunk 7 above)
Comment (timbussmann) [reply src/NServiceBus.Extensions.DependencyInjection.Autofac.AcceptanceTests/When_using_externally_managed_container.cs]: could make sense, I'll add that

File: src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
+﻿namespace NServiceBus
+{
+    using Extensions.DependencyInjection;
+    using Microsoft.Extensions.DependencyInjection;
+
+    /// <summary>
+    /// Provides factory methods for creating endpoints instances with an externally managed container implementing the Microsoft dependency injection abstraction.
```
Comment (timbussmann) [inline src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs]:
```suggestion
    /// Provides factory methods for creating endpoint instances with an externally managed container implementing the Microsoft dependency injection abstraction.
```

File: src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    /// </summary>
+    public static class EndpointWithExternallyManagedServiceProvider
+    {
+        /// <summary>
+        /// Creates a new startable endpoint based on the provided configuration that uses an externally managed container implementing Microsoft dependency injection abstraction.
```
Comment (timbussmann) [inline src/NServiceBus.Extensions.DependencyInjection/EndpointWithExternallyManagedServiceProvider.cs]:
```suggestion
        /// Creates a new startable endpoint based on the provided configuration that uses an externally managed container implementing the Microsoft dependency injection abstraction.
```

File: src/NServiceBus.Extensions.DependencyInjection/IStartableEndpointWithExternallyManagedContainerExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,26 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Extensions.DependencyInjection;
+
+    /// <summary>
+    /// An extension point for to create an endpoint in the start-up phase where the container is externally managed using the Microsoft dependency injection abstraction.
```
Comment (timbussmann) [inline src/NServiceBus.Extensions.DependencyInjection/IStartableEndpointWithExternallyManagedContainerExtensions.cs]:
```suggestion
    /// An extension point to create an endpoint in the start-up phase where the container is externally managed using the Microsoft dependency injection abstraction.
```

**Pull request merged**